### DISCUSSION TITLE
onSettled of useMutation getting called before onSettled of call-site mutate function

### DISCUSSION THREAD

**Discussion by KBrewz** [TanStack/query#4845] [Q&A] [answered]
I need the onSettled callback that gets passed into the call site of the mutate function to happen before the onSettled callback that is defined in the base useMutation hook. Is there a way to achieve this?

Currently they execute in the reverse order: base hook onSettled, then call-site onSettled.

**Comment by louis-young** [comment, marked as answer]
Hi 👋 

No, there's no way to achieve this. The callbacks passed to `useMutation` will always fire _before_ the callbacks passed to `mutate`. Be aware that there are also situations in which callbacks passed to `mutate` might not fire at all.

Please see: https://tkdodo.eu/blog/mastering-mutations-in-react-query#some-callbacks-might-not-fire

**Reply by louis-young** [reply]
The original question has been answered so I'm going to mark this as answered (I can't mark Dominik's elaboration as the answer as it's in reply to your suggested answer which is actually a question). 

If you need any more help, please start by explaining what exactly it is that you're trying to achieve instead of asking about an issue with your attempted solution. It's always useful to take a step back and look at the bigger picture when faced with an XY issue. If we know what you're trying to achieve then it'll be easier for us to help you 🙂

**Comment by KBrewz** [comment]
What is the reason for having the base-hook callbacks execute before the call-site callbacks?

If the call-site callbacks happened first, then it would allow the developer to choose the order by using `mutateAsync.then()` to execute functionality after the base-hook callbacks.

**Reply by TkDodo** [reply]
callbacks on useMutation always fire, and you can return a Promise from there and it will be awaited. You can't do that with the mutate callbacks.

You're asking for a very specific thing, but you haven't mentioned _why_ you need that (or think you need that). Often, there are other ways to achieve what you _actually_ want, but we can't help if you don't tell us the problem you're trying to solve.

You're pretty much asking a classic XY question - I've written about that topic here: https://tkdodo.eu/blog/how-can-i
